README.md
## What does Cowpy do?

Cowpy is a drop-in replacement for Python's `logging` with the additional benefits of:

- colorized logging
- easy setting of context/state
- intelligent auto-naming of loggers
- external, hierarchical logging configuration


### development 

* should we wrap the static loggers.. keep a handle on each logger requested and actually use the instance of Cowpy as the logging object in calling code?




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://github.com/tpalko/cowpy/compare/v0.0.3...v0.0.4) (2023-09-30)

### [0.0.3](https://github.com/tpalko/cowpy/compare/v0.0.2...v0.0.3) (2023-04-28)

### [0.0.2](https://github.com/tpalko/cowpy/compare/v0.0.1...v0.0.2) (2023-04-28)

### 0.0.1 (2023-03-09)


### Bug Fixes

* not loading a dictConfig if it is the one already loaded. fixed some dict handling in _log override. not fixing formatters for loggers if that logger-by-name has already been fixed ([b44134f](https://github.com/tpalko/cowpy/commit/b44134f4fa45e78d0fccb5159526cbc0911009cb))


tests/test.py
import sys
import os 
sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', 'src'))

import cowpy
logger = cowpy.getLogger()

logger.debug("edebug")
logger.warn('hey dooder')
logger.warning('this is  a warning')

class Testcowpy(object):

    logger = None 
    classlogger = cowpy.getLogger()

    def __init__(self, *args, **kwargs):
        self.logger = cowpy.getLogger()

    def doathing(self):
        self.logger.error('done a thing')
    
    def anotherthing(self):
        self.classlogger.info('wait, from a class level logger??')

import testsubfolder.blankconfig.blank

tcs = Testcowpy()
tcs.doathing()
tcs.anotherthing()

import testsubfolder.another 

tcs.doathing()
tcs.anotherthing()

logger.success('SUICCESSE!@ after all that!?')

logger.critical('error after all that!?')

tests/blankconfig/blank.py
import cowpy 

logger = cowpy.getLogger()

logger.debug('blank config test')

tests/testsubfolder/another.py
import cowpy 

logger = cowpy.getLogger()

logger.set_context('subbing')
logger.error('this is from a subfolder')
logger.clear_context()

def functionlog_declared_outside():
    logger.set_context('bare function')
    logger.warning('whoa there buddy, this logger ain\'t declared for the both of us!')

    fnLogger = cowpy.getLogger()
    fnLogger.info('but good thing, this one was created from within functionlog_declared_outside')

functionlog_declared_outside()

src/slate/slate.py
from enum import Enum 
import logging 
from logging.config import dictConfig #, DictConfigurator, dictConfigClass
import traceback 
import inspect 
import sys 
import os 
import json
import ast 

FOREGROUND_COLOR_PREFIX = '\033[38;2;'
FOREGROUND_COLOR_SUFFIX = 'm'
FOREGROUND_COLOR_RESET = '\033[0m'

COLOR_TABLE = {
    'white': '255;255;255',
    'red': '255;0;0',
    'darkred': '192;0;0',
    'green': '0;255;0',
    'orange': '255;165;0',
    'gray': '192;192;192',
    'darkgray': '128;128;128',
    'yellow': '165:165:0'
}

class Color(Enum):
    WHITE = COLOR_TABLE['white']
    RED = COLOR_TABLE['red']
    DARKRED = COLOR_TABLE['darkred']
    GREEN = COLOR_TABLE['green']
    ORANGE = COLOR_TABLE['orange']
    GRAY = COLOR_TABLE['gray']
    DARKGRAY = COLOR_TABLE['darkgray']
    YELLOW = COLOR_TABLE['yellow']

LEVEL_COLORS = {
    'DEBUG': Color.DARKGRAY,
    'INFO': Color.WHITE,
    'WARN': Color.ORANGE,
    'WARNING': Color.ORANGE,
    'SUCCESS': Color.GREEN,
    'ERROR': Color.RED,
    'CRITICAL': Color.DARKRED
}

LOG_LEVEL_DEFAULT = 'INFO'
LOG_LEVEL = os.getenv('LOG_LEVEL', LOG_LEVEL_DEFAULT)
LEVEL_COLORS = { logging._nameToLevel[k]: LEVEL_COLORS[k] for k in LEVEL_COLORS.keys() }
FORMATTER_BASE = '[ %(levelname)7s ] %(asctime)s %(name)s %(filename)12s:%(lineno)-4d %(message)s'

class SlateLogger(logging.Logger):
    
    # def __init__(self, *args, **kwargs):
    #     super(SlateLogger, self).__init__(*args, **kwargs)

    def _log(self, level, msg, args, exc_info=None, extra=None, stack_info=False, stacklevel=1):
        color_ext = { 'color': LEVEL_COLORS[level].value }
        extra = color_ext if not extra else extra.update(color_ext)
        # print(extra)
        # print([ h.formatter._fmt for h in self.handlers ])
        super(SlateLogger, self)._log(level, msg, args, exc_info=exc_info, extra=extra, stack_info=stack_info, stacklevel=stacklevel)
    
    def warn(self, msg):
        super(SlateLogger, self).warning(msg)
    
    def success(self, msg):
        super(SlateLogger, self).log(level=logging._nameToLevel['SUCCESS'], msg=msg)

class Slate(object):
    
    _intlogger = None 
    logger = None 
    quiet = False 
    headers = True 
    log_level = None 
    context = None 
    dry_run = False 

    dictConfigs = None 
    formatter_ids = None 
    
    def _getrc(self, path=None, logger=None):
        
        if not logger:
            logger = logging.getLogger()

        if not self.dictConfigs:
            self.dictConfigs = {}

        fullpath = os.path.realpath('.slaterc')

        if path:
            fullpath = os.path.join(path, '.slaterc')
        
        if fullpath not in self.dictConfigs:
            if os.path.exists(fullpath):            
                logger.info(f'{fullpath} exists. Opening, reading, and evaluating..')
                with open(fullpath, 'r') as conf:
                    try:
                        self.dictConfigs[fullpath] = ast.literal_eval(conf.read())
                    except:
                        logger.error(sys.exc_info()[0])
                        logger.error(sys.exc_info()[1])
                        traceback.print_tb(sys.exc_info()[2])

        if fullpath in self.dictConfigs:        
            logger.info(f'Calling dictConfig with {fullpath}')
            dictConfig(self.dictConfigs[fullpath])
            self.formatter_ids = []
        else:
            logger.warning(f'{fullpath} could not be found, opened, read, or evaluated')
    
    def _colorFormatter(self, fmt=FORMATTER_BASE):
        return logging.Formatter(fmt = f'{FOREGROUND_COLOR_PREFIX}%(color)s{FOREGROUND_COLOR_SUFFIX}{fmt}{FOREGROUND_COLOR_RESET}')

    def getLogger(self, name=None, internal=False):

        callingFrame = inspect.getouterframes(inspect.currentframe())[1]
        
        self._getrc(path=os.path.dirname(callingFrame.filename), logger=self._intlogger)

        # print(dir(callingFrame))
        name = name or callingFrame.function
        if name == "<module>":
            name = os.path.basename(callingFrame.filename)
        
        new_logger = logging.getLogger(name)
        if internal:
            self._intlogger = new_logger
        
        self._intlogger.info(f'Got -- {name} -- logger')
        
        self._intlogger.info(f'Fixing {len(new_logger.handlers)} handler formatters')
        for h in new_logger.handlers:
            self._intlogger.info(f'fixing formatters for handler {h}')
            if id(h.formatter) not in self.formatter_ids:
                h.setFormatter(self._colorFormatter(h.formatter._fmt))
                self.formatter_ids.append(id(h.formatter))
                self._intlogger.info(f'formatter {id(h.formatter)} processed..')
            else:
                self._intlogger.info(f'formatter {id(h.formatter)} already processed..')

        return new_logger
    
    def __init__(self, *args, **kwargs):

        self.getLogger(__name__, internal=True)
        self._intlogger.info(f'Slate internal logging configured as {__name__}')

    def clear_context(self):
        self.context = None 
        
    def set_context(self, context):
        self.context = context 
    
    def wrap_context(self, message):
        if self.dry_run:
            message = f'[ DRY RUN ] {message}'
        if self.context:
            message = f'[ {self.context} ] {message}'
        return message 
    
    # def _wrap(self, call, message, color=None):
    #     if not self.quiet:
    #         message = self.wrap_context(message)
    #         if color:
    #             call(colorwrapper(message, color))
    #         else:
    #             call(message)
    
    def text(self, message):
        if not self.quiet:
            self.logger.warning(message)
            
    # def debug(self, message):
    #     self._wrap(self.logger.debug, message)
    
    # def info(self, message):
    #     self._wrap(self.logger.info, message)
    
    # def warn(self, message):
    #     self._wrap(self.logger.warning, message)

    # def warning(self, message):
    #     self._wrap(self.logger.warning, message)

    # def success(self, message):
    #     self._wrap(self.logger.info, message)
    
    # def error(self, message):
    #     self._wrap(self.logger.error, message)

    def exception(self, data=False):
        stack_summary = traceback.extract_tb(sys.exc_info()[2])
        self.logger.error(stack_summary)
        if logging._nameToLevel[self.log_level.upper()] <= logging.ERROR:
            self.logger.error(sys.exc_info()[0])
            self.logger.error(sys.exc_info()[1])
            for line in stack_summary.format():
                self.logger.error(line)


src/cowpy/cowpy.py
from enum import Enum 
import logging 
import json
from logging.config import dictConfig
import traceback 
import inspect 
import sys 
import os 
import ast
from unittest.mock import DEFAULT 

FOREGROUND_COLOR_PREFIX = '\033[38;2;'
FOREGROUND_COLOR_SUFFIX = 'm'
FOREGROUND_COLOR_RESET = '\033[0m'
FORMATTER_BASE = '[ %(levelname)8s ] %(asctime)s %(filename)12s:%(lineno)5d %(name)12s %(context)-12s %(message)s'

COLOR_TABLE = {
    'white': '255;255;255',
    'red': '255;0;0',
    'darkred': '192;0;0',
    'green': '0;255;0',
    'orange': '255;165;0',
    'gray': '192;192;192',
    'darkgray': '128;128;128',
    'yellow': '165:165:0'
}

class Color(Enum):
    WHITE = COLOR_TABLE['white']
    RED = COLOR_TABLE['red']
    DARKRED = COLOR_TABLE['darkred']
    GREEN = COLOR_TABLE['green']
    ORANGE = COLOR_TABLE['orange']
    GRAY = COLOR_TABLE['gray']
    DARKGRAY = COLOR_TABLE['darkgray']
    YELLOW = COLOR_TABLE['yellow']

LEVEL_COLORS = {
    'DEBUG': Color.DARKGRAY,
    'INFO': Color.WHITE,
    'WARN': Color.ORANGE,
    'WARNING': Color.ORANGE,
    'SUCCESS': Color.GREEN,
    'ERROR': Color.RED,
    'CRITICAL': Color.DARKRED
}

LEVEL_COLORS = { logging._nameToLevel[k]: LEVEL_COLORS[k] for k in LEVEL_COLORS.keys() }

DEFAULT_RC_PATH = '__DEFAULT_RC_PATH__'

class CowpyConfigurator():

    _config = None 

    def __init__(self, *args, **kwargs):
        self._config = args[0]

    def configure(self):
        pass 

class CowpyLogger(logging.Logger):

    context = None 

    # -- overriding base 
    def _log(self, level, msg, args, exc_info=None, extra=None, stack_info=False, stacklevel=1):
        if level not in LEVEL_COLORS:
            raise NotImplementedError(f'There is no {level} in LEVEL_COLORS. Go ahead and check it.')

        extra = extra or {}

        extra.update({ 
            'color': LEVEL_COLORS[level].value,
            'context': f'[ {self.context or "-"} ]'
        })

        # padded = ''
        # context = True
        # if self.context:
        #     for h in self.handlers:
        #         if h.formatter and h.formatter._fmt.find('context') < 0:
        #             context = False 
        #             break 
        #     if context:
        #         extra.update({'context': f'[ {self.context} ]'})
        #     padded = self.context[0:8]
        # while len(padded) < 8:
        #     padded = f'{padded} '
        # msg = f'[ {padded} ] {msg}'
        
        super(CowpyLogger, self)._log(level, msg, args, exc_info=exc_info, extra=extra, stack_info=stack_info, stacklevel=stacklevel)
    
    def warn(self, msg):
        super(CowpyLogger, self).warning(msg)
    
    def success(self, msg):
        super(CowpyLogger, self).log(level=logging._nameToLevel['SUCCESS'], msg=msg)

    def clear_context(self):
        self.context = None 
        
    def set_context(self, context):
        self.context = context 
    
    def exception(self):
        stack_summary = traceback.extract_tb(sys.exc_info()[2])
        full_error_text = f'{sys.exc_info()[1]}: {sys.exc_info()[0]} {"".join(stack_summary.format())}'
        self.error(full_error_text)
        # self.error(sys.exc_info()[0])
        # self.error(sys.exc_info()[1])
        # self.error("".join(stack_summary.format()))
        # self.error("\n".join([ f'{s.filename} line {s.line}' for s in stack_summary ]))
        # for line in stack_summary.format():
        #     self.error(line)
        # if logging._nameToLevel[self.log_level.upper()] <= logging.ERROR:
        #     self.error(sys.exc_info()[0])
        #     self.error(sys.exc_info()[1])
        #     for line in stack_summary.format():
        #         self.error(line)
  

class Cowpy(object):
    
    _intlogger = None 
    
    dictConfigs = None 
    _loaded_dict_config_path = None 
    _fixed_logger_names = [] 
    formatter_ids = None 
    
    context_enabled = False 
    
    def __init__(self, *args, **kwargs):

        # -- set internal logger 
        self._intlogger = self.getLogger(__name__)

        self._log_internal('debug', f'Cowpy internal logging configured as {__name__}')

        for k in kwargs:
            try:
                self.__getattribute__(k)
                self.__setattr__(k, kwargs[k])
            except:
                self._log_internal('error', f'Cowpy has no {k}')
    
    def _get_default_config(self, logger_name, level='WARNING'):

        return {
            'version': 1,
            'disable_existing_loggers': False,
            'formatters': {
                'default': {
                    'format': FORMATTER_BASE
                },
            },
            'handlers': {
                'console': {
                    'class': 'logging.StreamHandler',
                    'formatter': 'default',
                    'level': f'{level}'
                },
            },
            'loggers': {    
                f'{logger_name}': {
                    'handlers': ['console'],
                    'level': f'{level}',
                    'propagate': False,
                },
            }
        }

    def _getrc_path(self, calling_filename):
        
        self._log_internal_context('cowpy rc path')

        if not self.dictConfigs:
            self.dictConfigs = {}

        # -- realpath will resolve symlinked calling code 
        calling_folder = os.path.dirname(os.path.realpath(calling_filename))
        self._log_internal('debug', f'Using {calling_folder} as calling folder')
        rc_folder = calling_folder 
        final_rc_path = None 

        # -- follow parent folders all the way up, searching desperately for configuration
        while not os.path.exists(os.path.join(rc_folder, '.cowpyrc')):
            if rc_folder == '/':
                self._log_internal('debug', f'Reached the root folder with no RC')
                rc_folder = None 
                break 
            self._log_internal('debug', f'RC not found in rc_folder, going up one..')
            rc_folder = os.path.realpath(os.path.join(rc_folder, os.pardir))

        if rc_folder:
            final_rc_path = os.path.join(rc_folder, '.cowpyrc')

        # -- we already checked above, but double check 
        if final_rc_path and not os.path.exists(final_rc_path):
            final_rc_path = None 
        
        if not final_rc_path:
            self._log_internal('warning', f'.cowpyrc in {calling_folder} or its parents could not be found, opened, read, or evaluated')
            self._log_internal('warning', f'Calling dictConfig with default config')  
            final_rc_path = DEFAULT_RC_PATH
        
        self._log_internal_context(None)

        return final_rc_path
    
    def _fix_config_formatters(self, config_dict):
        formatters = config_dict['formatters'] if 'formatters' in config_dict else {}                        
        for handler_name in config_dict['handlers'].keys():
            handler = config_dict['handlers'][handler_name]
            if handler['class'] != 'logging.StreamHandler':
                continue 
            if 'formatter' in handler:
                formatter_name = handler['formatter']
                if formatter_name not in formatters:
                    self._log_internal('debug', f'Adding a base formatter for handler {handler_name} {formatter_name}')
                    formatters[formatter_name] = { 'format': FORMATTER_BASE }
                else:
                    self._log_internal('debug', f'Found formatter {formatters[formatter_name]} for handler {handler_name}')
            else:
                self._log_internal('debug', f'Setting handler {handler_name} formatter to "default"')
                handler['formatter'] = 'default'
                if 'default' not in formatters:
                    self._log_internal('debug', f'Adding missing "default" formatter')
                    formatters['default'] = { 'format': FORMATTER_BASE }
        config_dict['formatters'] = formatters
        # return config_dict

    def _load_rc_contents(self, rc_path, logger_name):

        rcFileContents = None 

        if os.path.exists(rc_path):
            if rc_path not in self.dictConfigs:            
                self._log_internal('debug', f'{rc_path} exists. Opening, reading, and evaluating..')
                with open(rc_path, 'r') as conf:
                    try:
                        config_dict = ast.literal_eval(conf.read())
                        self._log_internal('debug', json.dumps(config_dict, indent=4))
                        self._fix_config_formatters(config_dict)
                        self.dictConfigs[rc_path] = config_dict
                    except:
                        self._log_internal('error', sys.exc_info()[0])
                        self._log_internal('error', sys.exc_info()[1])
                        traceback.print_tb(sys.exc_info()[2])
            
            rcFileContents = self.dictConfigs[rc_path]            

        if not rcFileContents:

            if rc_path != DEFAULT_RC_PATH:
                self._log_internal('warning', f'RC not found from {rc_path}, getting default config')

            rcFileContents = self._get_default_config(logger_name=logger_name)

        return rcFileContents
    
    def _colorFormatter(self, fmt=FORMATTER_BASE):
        return logging.Formatter(fmt = f'{FOREGROUND_COLOR_PREFIX}%(color)s{FOREGROUND_COLOR_SUFFIX}{fmt}{FOREGROUND_COLOR_RESET}')

    def _handler(self):
        newH = logging.StreamHandler()        
        newH.setFormatter(self._colorFormatter())
        return newH 

    def _log_internal(self, level, msg):
        
        _local_logger = self._intlogger
        if not _local_logger:
            _local_logger = logging.getLogger(__name__)
            
        _local_logger.log(level=logging._nameToLevel[level.upper()], msg=msg, stacklevel=3)

    def _log_internal_context(self, context):
        if self._intlogger:
            self._intlogger.set_context(context)

    def fixLoggerFormatters(self, logger_name):

        if logger_name in self._fixed_logger_names:
            return 

        logger = logging.getLogger(logger_name)
        self._log_internal('debug', f'Fixing formatters for {len(logger.handlers)} {logger_name} handlers: {",".join([ h.__class__.__name__ for h in logger.handlers ])}')
        if len(logger.handlers) > 0:
            for h in logger.handlers:
                if h.formatter:
                    # self._log_internal('debug', dir(h.formatter))
                    if id(h.formatter) not in self.formatter_ids:
                        self._log_internal('debug', f'Replacing {h.__class__.__name__} formatter')
                        newF = self._colorFormatter(h.formatter._fmt)
                        self.formatter_ids.append(id(h.formatter))                        
                        h.setFormatter(newF)
                    else:
                        self._log_internal('debug', f'Doing nothing for {h.__class__.__name__} formatter (filed as done)')
                else:
                    self._log_internal('debug', f'Adding formatter for {h.__class__.__name__}')
                    h.setFormatter(self._colorFormatter())
        else:
            self._log_internal('debug', f'No handlers on this new logger, so adding a default')
            logger.addHandler(self._handler())      
        
        self._fixed_logger_names.append(logger_name)

    def getLogger(self, name=None, config=None):
        
        self._log_internal('debug', f'Someone is asking for a logger ({name})!')
        callingFrame = inspect.getouterframes(inspect.currentframe())[1]
        
        self._log_internal('debug', inspect.getouterframes(inspect.currentframe())[1].filename)
        # self._log_internal('debug', f'{dir(callingFrame.frame)}')
        # self._log_internal('debug', f'{dir(callingFrame.frame.f_code)}')
        # self._log_internal('debug', f'{callingFrame.frame.f_code.co_names}')

        # self._log_internal('debug', f'have calling frame: {callingFrame.function}, fallback is {name}')
        # print(dir(callingFrame))
        if not name:            
            name = callingFrame.function
            self._log_internal('debug', f'Logger name changed to {name}')

        if name == "<module>":            
            name =  os.path.splitext(os.path.basename(callingFrame.filename))[0]
            self._log_internal('debug', f'Logger name changed to {name}')
            # self._log_internal('debug', f'<module> is no good, so going for calling frame filename: {name}')

        self._log_internal('debug', f'vvvvvvvvvvvvvvvvvvv getting logger: {name} vvvvvvvvvvvvvvvvvvvvvvvvvv')

        rc_path = config or self._getrc_path(calling_filename=callingFrame.filename)
        
        rcFileContents = None 

        if self._loaded_dict_config_path != rc_path:

            self._log_internal('debug', f'loading contents of RC file {rc_path}')
            rcFileContents = self._load_rc_contents(rc_path, name)      

            self._log_internal('debug', f'Calling dictConfig with')
            self._log_internal('debug', json.dumps(rcFileContents, indent=4))
                
            dictConfig(rcFileContents)    
        
            self._loaded_dict_config_path = rc_path

            # -- applying a configuration invalidates any ids that might have been in memory
            # -- so we dump our tracking of those IDs
            self.formatter_ids = []
            self._fixed_logger_names = []
            
            

            other_loggers = [ n for n in rcFileContents['loggers'].keys() if n != name ]
            # self._log_internal('info', f'Fixing logger formatters for {len(other_loggers)} configured loggers (other than {name}): { ",".join(other_loggers)}')
            for logger_name in other_loggers:
                # -- ignore the root logger, this is a bad time 
                if logger_name != '':
                    _ = self.fixLoggerFormatters(logger_name)
            
            self.fixLoggerFormatters(name)

        else:
            self._log_internal('debug', f'RC found is what is already loaded')            

        self._log_internal('debug', f'^^^^^^^^^^^^^^^^^^ getting logger: {name} ^^^^^^^^^^^^^^^^^^^^^^^^')

        named_logger = logging.getLogger(name)
        self._log_internal('info', f'Got {name} logger')
        return named_logger
